Test data stream downsample

salvatore-campagna opened the pull request:
Test a data stream including a downsampled index.

salvatore-campagna committed:
test: data stream downsample

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

martijnvg commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java:
```diff
@@ -0,0 +1,137 @@
+        indexDocs(dataStream.getName(), 10);
+
+        // Rollover
+        final DataStream newDataStream = dataStream.rollover(
```
This should be done in the rollover api? This isn't stored and doesn't perform the entire rollover operation (also creating a new index).

salvatore-campagna committed:
test: use the data stream api

salvatore-campagna committed:
test: execute a date histogram query on a data stream

The data stream is created so that it includes a downsampled index and
a normal write index. The date histogram query uses a fixed_interval value
that is smaller than the downsample fixed_interval aggregation interval.

salvatore-campagna commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java:
```diff
@@ -0,0 +1,225 @@
+        // future.
+        // 10_000_000 milliseconds is a bit more than 3 hours, which is more than enough to make sure we are not writing documents to the
+        // old data stream index.
+        indexDocs(dataStreamName, 10, A_BIT_MORE_THAN_3_HOURS_MILLIS);
```
maybe there is a way to programmatically get the `start_time` of the new backing index? Will check that.

martijnvg replied:
Yes, maybe use the get index settings api to get the start_time setting? And then just increment 1 ms for each document to be indexed?

martijnvg replied:
So instead just pass the start time to `indexDocs(...)` method? And then iteratively add a ms to the start time and use that as timestamp for the document to be indexed?

salvatore-campagna commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java:
```diff
@@ -0,0 +1,225 @@
+        // date histogram. We know, anyway, that we will have 10 documents in the first bucket, 10 documents in the last bucket
+        // and a variable number of intermediate buckets with exactly 0 documents. This is a result of the way downsampling deals
+        // with a fixed interval granularity that is larger than the date histogram fixed interval (1 minute (date histogram fixed_interval)
+        // < 1 hour (downsample fixed_interval)).
```
Yes, here I am assuming that indexing a batch of 10 documents takes less than a minute...which I believe is a fair assumption...

martijnvg replied:
Can we run this search request before downsampling and then compare after downsampling that the results are the same as the results of the first search request?

salvatore-campagna replied:
The original index (the one used as source index while doing downsampling) is removed from the data stream. That is because, at the moment, we still don't support multiple indices with the same time span, and we miss the logic to select the right index to use for each query.

martijnvg replied:
👍

nik9000 replied:
It looks like you are using a fixed timestamp for the generated documents. It looks like you are assuming that the 10 documents don't straddle the line. Maybe you should detect this and round up to the start of the next bucket or something. Or just be ok if they straddle in the assertion.

salvatore-campagna replied:
I think I will test for something like this

```
sum(first and second bucket doc count) == 10 &&
sum(last two buckets doc count) == 0
sum(whatever is in between doc count) == 0
```

martijnvg commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java:
```diff
@@ -0,0 +1,225 @@
+import static org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.DEFAULT_TIMESTAMP_FIELD;
+import static org.hamcrest.Matchers.equalTo;
+
+public class DataStreamTests extends ESIntegTestCase {
```
Can this extend from `ESSingleNodeTestCase`? I don't think there is a need to run setup a cluster with multiple nodes?

martijnvg commented on x-pack/plugin/rollup/build.gradle:
```diff
@@ -14,6 +14,7 @@ dependencies {
   compileOnly project(path: xpackModule('ilm'))
   compileOnly project(path: xpackModule('mapper-aggregate-metric'))
   testImplementation(testArtifact(project(xpackModule('core'))))
+  testImplementation project(':modules:lang-painless')
```
Just double checking, what part of the test requires the usage of painless?

salvatore-campagna replied:
Sorry, this was a left-over...

salvatore-campagna committed:
test: improve test

salvatore-campagna committed:
test: remove unused constant

salvatore-campagna committed:
fix: remove unused dependency

salvatore-campagna committed:
fix: extend ESSingleNodeTestCase

salvatore-campagna committed:
fix: use now as the start time

salvatore-campagna committed:
Merge branch 'main' into test/data-stream-downsample

martijnvg commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java:
```diff
@@ -0,0 +1,235 @@
+            .stream()
+            .map(Index::getName)
+            .toList();
+        assertThat(backingIndices, Matchers.containsInAnyOrder(downsampleTargetIndex, rolloverResponse.getNewIndex()));
```
Internally the backing indices are stored as a list in cluster state and I think are returned as an array. So we can assume that the indices are returned in consistent order. So I think you can just use `contains()` here.

martijnvg commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java:
```diff
@@ -0,0 +1,235 @@
+            .execute(DownsampleAction.INSTANCE, downsampleRequest)
+            .actionGet();
+
+        client().admin()
```
Maybe add this is a comment?

salvatore-campagna replied:
I was actually trying to see if I can fix the issue...I will spent a couple more hours if I can fix it (maybe, even as part of another review so that I can ask for someone else's review).

martijnvg reviewed (approved):
LGTM

salvatore-campagna committed:
fix: replace containsInAnyOrder with contains

salvatore-campagna committed:
comment: possible bug while checking for settings/metadata updates

salvatore-campagna committed:
nit: remove useless braces

salvatore-campagna commented:
@elasticsearchmachine test this please

csoulios reviewed (commented):
I do not quite understand what this test is testing. I think there are no clean boundaries between setting up the test and doing the actual test. Until line 144, the test reproduces the steps performed by ILM when downsampling a data stream. I think this functionalty has been tested by tests such as [`DownsampleActionSingleNodeTests.testRollupDatastream()`](https://github.com/elastic/elasticsearch/blob/aa95070ea3e7ac41485ecc7aead542b21b035402/x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DownsampleActionSingleNodeTests.java#L542-L572) and [`DownsampleActionIT.testRollupDatastream()`](https://github.com/elastic/elasticsearch/blob/aa95070ea3e7ac41485ecc7aead542b21b035402/x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DownsampleActionSingleNodeTests.java#L542-L572)

One of the tests I think we should include is  to run some aggregations before downsampling a datastream, then downsample one of its backing indices  and then run the same aggs again and ensure we get the same results.

nik9000 commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java:
```diff
@@ -0,0 +1,240 @@
+            .execute(DownsampleAction.INSTANCE, downsampleRequest)
+            .actionGet();
+
+        /**
```
It's just `/*` instead of `/**` because this isn't javadoc.

salvatore-campagna commented:
> 

Hi the the first test is testing that:

"the rollup index has not been added to the data stream" and that "the source index is still a member of the data stream"

while the second one is testing that rollup happens when transitioning from one ILM phase to the next.

The test I wrote checks what happens if we query a data stream where one index is downsampled and the other not, after replacing one of the indices (the rollover old index) with a downsampled one.

This is the reason why I check for the 10 documents in the first and last buckets with "intermediate" empty buckets.

salvatore-campagna committed:
nit: not a javadoc comment

salvatore-campagna committed:
fix: handle possible generation of documents on a minute boundary

csoulios commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java:
```diff
@@ -0,0 +1,248 @@
+import static org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.DEFAULT_TIMESTAMP_FIELD;
+import static org.hamcrest.Matchers.equalTo;
+
+public class DataStreamTests extends ESSingleNodeTestCase {
```
Does it make sense to rename the class to `DownsampleDatastreamTests`? I know the test is included in the `org.elasticsearch.xpack.downsample` package so adding "downsample" to the test name feels unnecessary. I was only thinking it would help with IDE autocomplete and search features. 

This is merely a suggestion, so feel free to ignore it.

csoulios commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java:
```diff
@@ -0,0 +1,248 @@
+        client().execute(ModifyDataStreamsAction.INSTANCE, modifyDataStreamRequest).actionGet();
+
+        // THEN
+        assertThat(downsampleResponse.isAcknowledged(), equalTo(true));
```
Maybe you should move this assertion immediately after the downsample action is invoked. Here, it feels a bit off place. 

Alternatively, you can drop this assertion as none of the other actions are being asserted.

salvatore-campagna replied:
I placed it there since it is after the "THEN" because i one of the things to verify. When I write such tests I always try to separate the "GIVEN", "WHEN" and "THEN" sections in such a way to have assertions all grouped in the "THEN" section, avoiding mixing actions with the verification of the outcome.

csoulios reviewed (approved):
LGTM!

I added some minor comments

salvatore-campagna committed:
nit: change class name

salvatore-campagna commented:
@elasticsearchmachine test this please

salvatore-campagna commented:
@elasticsearchmachine test this please

salvatore-campagna committed:
Merge branch 'main' into test/data-stream-downsample

salvatore-campagna commented:
@elasticsearchmachine test this please
